content/en/blog/_posts/2021-09-06-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,116 @@
+
+![](./memory-qos-cal.svg)
+
+The following table details the specific functions of these two parameters and how they correspond to Kubernetes container resources.
+
+<table>
+    <tr>
+        <th style="text-align:center">File</th>
+        <th style="text-align:center">Description</th>
+   </tr>
+   <tr>
+        <td>memory.min</td>
+        <td><code>memory.min</code> specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked.
+        <br>
+        <br>
+        <i>We map it to <code>requests.memory</code></i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. 
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on <code>limits.memory</code> or <code>node allocatable memory</code> and a  throttling factor</i>
+       </td>
+   </tr>
+</table>
+
+When container memory requests are made, kubelet passes `memory.min` to the back-end CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during container creation. The `memory.min` in container level cgroup will be set to:  
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all set, the pod and node cgroup directories need to be set rightly. 
+
+`memory.min` in pod level cgroup:  
+![](./pod-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` in node level cgroup:  
+![](./node-memory-min.svg)  
+<sub>i: the i<sup>th</sup> pod in one node, j: the j<sup>th</sup> container in one pod</sub>
+
+However, unlike container runtimes, kubelet communicates with cgroup v2 directly through libcontainer of runc for pod and node level cgroups.
+
+For memory limits, in addition to the original way of limiting memory usage, Memory QoS adds an additional feature of throttling memory allocation. A throttling factor is introduced as a multiplier (default is 0.8). If the result of multiplying memory limits by the factor is greater than memory requests, kubelet will set `memory.high` to the value and use `Unified` via CRI. And if the container does not specify memory limits, kubelet will use node allocatable memory instead. The `memory.high` in container level cgroup is set to:
```

Comment:
I think this does a great job explaining how this works but there's not a lot of details of the end user impact:

I would add a sentence on the end user impact, i.e. something like

This can can help improve stability when pod memory usage increases, ensuring that memory is throttled as it approaches the memory limit.

What do you think?